File: daily-coding-challenge/2026/august/py/bucket_fill3.py
"""

The Last Challenge: Bucket Fill 3

Today marks a year of daily coding challenges. This is the last new one for now. Good luck!

Given a 2D grid of single-letter color strings and a target color, return the minimum number of flood fill "clicks" needed to make the entire grid that color.
  Each click changes the clicked cell's color and the entire region of connected cells of the same color (4-directional).
  Clicks can use any color as an intermediate step, not just the target color.

Tests:
  bucket_fill([["B", "B"], ["B", "B"]], "R") should return 1.
  bucket_fill([["G", "G", "G"], ["G", "G", "G"], ["G", "G", "G"]], "G") should return 0.
  bucket_fill([["P", "P", "Y"], ["Y", "P", "Y"], ["Y", "P", "P"]], "O") should return 2.
  bucket_fill([["G", "Y", "C", "C"], ["Y", "Y", "Y", "B"], ["C", "Y", "B", "B"], ["C", "B", "B", "C"]], "R") should return 4.
  bucket_fill([["G", "G", "O", "O"], ["G", "Y", "B", "Y"], ["B", "Y", "B", "Y"], ["B", "Y", "B", "Y"], ["G", "G", "G", "G"]], "P") should return 5.
  bucket_fill([["R", "G", "R", "G"], ["R", "G", "R", "G"], ["B", "B", "B", "B"], ["B", "B", "B", "B"], ["R", "G", "R", "G"]], "Y") should return 3.

"""

from collections import deque

def bucket_fill(grid, target_color):

    R = len(grid)
    C = len(grid[0])
    
    initial_state = tuple(tuple(row) for row in grid)
    
    if all(cell == target_color for row in initial_state for cell in row):
        return 0
        
    queue = deque([(initial_state, 0)])
    visited = {initial_state}
    
    colors = set(target_color)
    for row in initial_state:
        for cell in row:
            colors.add(cell)
            
    while queue:
        state, steps = queue.popleft()
        
        if all(cell == target_color for row in state for cell in row):
            return steps
            
        visited_cells = set()
        components = []
        
        for r in range(R):
            for c in range(C):
                if (r, c) not in visited_cells:
                    color = state[r][c]
                    comp = []
                    stack = [(r, c)]
                    visited_cells.add((r, c))
                    
                    while stack:
                        cr, cc = stack.pop()
                        comp.append((cr, cc))
                        for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                            nr, nc = cr + dr, cc + dc
                            if 0 <= nr < R and 0 <= nc < C and (nr, nc) not in visited_cells and state[nr][nc] == color:
                                visited_cells.add((nr, nc))
                                stack.append((nr, nc))
                    components.append((color, comp))
                    
        for color, comp in components:
            for new_color in colors:
                if new_color == color:
                    continue
                
                new_state = [list(row) for row in state]
                for r, c in comp:
                    new_state[r][c] = new_color
                new_state_tuple = tuple(tuple(row) for row in new_state)
                
                if new_state_tuple not in visited:
                    visited.add(new_state_tuple)
                    queue.append((new_state_tuple, steps + 1))
                    
    return -1


print(
    bucket_fill(
        [
            ["B", "B"],
            ["B", "B"],
        ],
        "R",
    ),
)  # 1
print(
    bucket_fill(
        [
            ["G", "G", "G"],
            ["G", "G", "G"],
            ["G", "G", "G"],
        ],
        "G",
    ),
)  # 0
print(
    bucket_fill(
        [
            ["P", "P", "Y"],
            ["Y", "P", "Y"],
            ["Y", "P", "P"],
        ],
        "O",
    ),
)  # 2
print(
    bucket_fill(
        [
            ["G", "Y", "C", "C"],
            ["Y", "Y", "Y", "B"],
            ["C", "Y", "B", "B"],
            ["C", "B", "B", "C"],
        ],
        "R",
    ),
)  # 4
print(
    bucket_fill(
        [
            ["G", "G", "O", "O"],
            ["G", "Y", "B", "Y"],
            ["B", "Y", "B", "Y"],
            ["B", "Y", "B", "Y"],
            ["G", "G", "G", "G"],
        ],
        "P",
    ),
)  # 5
print(
    bucket_fill(
        [
            ["R", "G", "R", "G"],
            ["R", "G", "R", "G"],
            ["B", "B", "B", "B"],
            ["B", "B", "B", "B"],
            ["R", "G", "R", "G"],
        ],
        "Y",
    ),
)  # 3

File: daily-coding-challenge/2026/august/js/bucketFill3.js
/*

The Last Challenge: Bucket Fill 3

Today marks a year of daily coding challenges. This is the last new one for now. Good luck!

Given a 2D grid of single-letter color strings and a target color, return the minimum number of flood fill "clicks" needed to make the entire grid that color.
  Each click changes the clicked cell's color and the entire region of connected cells of the same color (4-directional).
  Clicks can use any color as an intermediate step, not just the target color.

Tests:
  bucketFill([["B", "B"], ["B", "B"]], "R") should return 1.
  bucketFill([["G", "G", "G"], ["G", "G", "G"], ["G", "G", "G"]], "G") should return 0.
  bucketFill([["P", "P", "Y"], ["Y", "P", "Y"], ["Y", "P", "P"]], "O") should return 2.
  bucketFill([["G", "Y", "C", "C"], ["Y", "Y", "Y", "B"], ["C", "Y", "B", "B"], ["C", "B", "B", "C"]], "R") should return 4.
  bucketFill([["G", "G", "O", "O"], ["G", "Y", "B", "Y"], ["B", "Y", "B", "Y"], ["B", "Y", "B", "Y"], ["G", "G", "G", "G"]], "P") should return 5.
  bucketFill([["R", "G", "R", "G"], ["R", "G", "R", "G"], ["B", "B", "B", "B"], ["B", "B", "B", "B"], ["R", "G", "R", "G"]], "Y") should return 3.

*/

function bucketFill(grid, targetColor) {
    const R = grid.length;
    const C = grid[0].length;

    function serialize(g) {
        return g.map(row => row.join('')).join('|');
    }

    const initialState = grid.map(row => [...row]);
    
    let allTarget = true;
    for (let r = 0; r < R; r++) {
        for (let c = 0; c < C; c++) {
            if (grid[r][c] !== targetColor) allTarget = false;
        }
    }
    if (allTarget) return 0;

    const colors = new Set([targetColor]);
    for (let r = 0; r < R; r++) {
        for (let c = 0; c < C; c++) {
            colors.add(grid[r][c]);
        }
    }

    const queue = [[initialState, 0]];
    const visited = new Set([serialize(initialState)]);

    while (queue.length > 0) {
        const [state, steps] = queue.shift();

        let matches = true;
        for (let r = 0; r < R; r++) {
            for (let c = 0; c < C; c++) {
                if (state[r][c] !== targetColor) {
                    matches = false;
                    break;
                }
            }
            if (!matches) break;
        }
        if (matches) return steps;

        const visitedCells = Array.from({ length: R }, () => Array(C).fill(false));
        const components = [];

        for (let r = 0; r < R; r++) {
            for (let c = 0; c < C; c++) {
                if (!visitedCells[r][c]) {
                    const color = state[r][c];
                    const comp = [];
                    const stack = [[r, c]];
                    visitedCells[r][c] = true;

                    while (stack.length > 0) {
                        const [cr, cc] = stack.pop();
                        comp.push([cr, cc]);

                        const dirs = [[-1, 0], [1, 0], [0, -1], [0, 1]];
                        for (const [dr, dc] of dirs) {
                            const nr = cr + dr;
                            const nc = cc + dc;
                            if (nr >= 0 && nr < R && nc >= 0 && nc < C && !visitedCells[nr][nc] && state[nr][nc] === color) {
                                visitedCells[nr][nc] = true;
                                stack.push([nr, nc]);
                            }
                        }
                    }
                    components.push({ color, comp });
                }
            }
        }

        for (const { color, comp } of components) {
            for (const newColor of colors) {
                if (newColor === color) continue;

                const newState = state.map(row => [...row]);
                for (const [r, c] of comp) {
                    newState[r][c] = newColor;
                }

                const serialized = serialize(newState);
                if (!visited.has(serialized)) {
                    visited.add(serialized);
                    queue.push([newState, steps + 1]);
                }
            }
        }
    }

    return -1;

}

console.log(
  bucketFill(
    [
      ["B", "B"],
      ["B", "B"],
    ],
    "R",
  ),
); // 1
console.log(
  bucketFill(
    [
      ["G", "G", "G"],
      ["G", "G", "G"],
      ["G", "G", "G"],
    ],
    "G",
  ),
); // 0
console.log(
  bucketFill(
    [
      ["P", "P", "Y"],
      ["Y", "P", "Y"],
      ["Y", "P", "P"],
    ],
    "O",
  ),
); // 2
console.log(
  bucketFill(
    [
      ["G", "Y", "C", "C"],
      ["Y", "Y", "Y", "B"],
      ["C", "Y", "B", "B"],
      ["C", "B", "B", "C"],
    ],
    "R",
  ),
); // 4
console.log(
  bucketFill(
    [
      ["G", "G", "O", "O"],
      ["G", "Y", "B", "Y"],
      ["B", "Y", "B", "Y"],
      ["B", "Y", "B", "Y"],
      ["G", "G", "G", "G"],
    ],
    "P",
  ),
); // 5
console.log(
  bucketFill(
    [
      ["R", "G", "R", "G"],
      ["R", "G", "R", "G"],
      ["B", "B", "B", "B"],
      ["B", "B", "B", "B"],
      ["R", "G", "R", "G"],
    ],
    "Y",
  ),
); // 3
